File: sdk/ruby/lib/phantex/hooks.rb
# frozen_string_literal: true

require_relative "hooks/base"
require_relative "hooks/ruby_openai"
require_relative "hooks/langchainrb"

module Phantex
  # Hook registry — maps hook name to class.
  HOOK_REGISTRY = {
    "ruby_openai"  => Hooks::RubyOpenAI,
    "langchainrb"  => Hooks::LangchainRb,
  }.freeze

  def self.install_hooks(config, transport)
    hooks_config = config.hooks.downcase.strip
    return [] if hooks_config == "none"

    names = if hooks_config == "auto"
              HOOK_REGISTRY.keys
            else
              hooks_config.split(",").map(&:strip)
            end

    installed = []
    names.each do |name|
      klass = HOOK_REGISTRY[name]
      next unless klass

      hook = klass.new(transport: transport, config: config)
      if hook.install
        installed << hook
      end
    rescue StandardError => e
      warn "[phantex] failed to install hook #{name}: #{e.message}" if config.debug
    end
    installed
  end
end

File: sdk/go/hooks.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025-2026 The Phantex Authors

package phantex

import (
	"log"
	"strings"
)

// Hook is the interface all framework hooks implement.
type Hook interface {
	Name() string
	Install() bool
	Uninstall()
}

// hookRegistry maps hook names to constructor functions.
var hookRegistry = map[string]func(Transport, *Config) Hook{
	"openai": newOpenAIHook,
	"http":   newHTTPHook,
}

// installHooks detects and installs configured hooks.
func installHooks(transport Transport, cfg *Config) []Hook {
	hooksConfig := strings.ToLower(strings.TrimSpace(cfg.Hooks))

	var names []string
	switch hooksConfig {
	case "none":
		return nil
	case "auto":
		for name := range hookRegistry {
			names = append(names, name)
		}
	default:
		for _, name := range strings.Split(hooksConfig, ",") {
			name = strings.TrimSpace(name)
			if name != "" {
				names = append(names, name)
			}
		}
	}

	var installed []Hook
	for _, name := range names {
		ctor, ok := hookRegistry[name]
		if !ok {
			if cfg.Debug {
				log.Printf("phantex: unknown hook: %s", name)
			}
			continue
		}
		hook := ctor(transport, cfg)
		func() {
			defer func() {
				if r := recover(); r != nil {
					if cfg.Debug {
						log.Printf("phantex: hook '%s' panicked during install: %v", name, r)
					}
				}
			}()
			if hook.Install() {
				installed = append(installed, hook)
				if cfg.Debug {
					log.Printf("phantex: hook '%s' installed", name)
				}
			}
		}()
	}
	return installed
}
